go/marshal/encode.go

```diff
@@ -447,18 +448,21 @@ func typeFields(t reflect.Type, parentStructTypes []reflect.Type, options encode
 			index:     i,
 			nomsType:  nt,
 			omitEmpty: tags.omitEmpty,
+			optional:  tags.omitEmpty,
```

Comment:
Let's talk about it today. It is clear that omitempty in MarshalType MUST imply optional because the field is not always there.